File: app/src/main/java/com/lunesu/pengchauferry/HolidayRepository.kt
package com.lunesu.pengchauferry

import org.joda.time.LocalDate

open class HolidayRepository(db: DbOpenHelper) {
    companion object {
        private val BUDDHA = LocalDate(HongKongHolidayFetcher.YEAR, 4, 30)
    }

    private val holidaysDao = HolidayDao(db)

    open fun getHoliday(day: LocalDate): Boolean {
        return holidaysDao.query(day)
    }

    open fun setHoliday(day: LocalDate, isHoliday: Boolean) {
        if (isHoliday) {
            holidaysDao.insert(day)
        } else {
            holidaysDao.delete(day)
        }
    }

    open fun shouldRefresh(): Boolean {
        return !getHoliday(BUDDHA)
    }

    open suspend fun refresh() {
        runCatching {
            holidaysDao.save(Utils.atLeast(HongKongHolidayFetcher.fetch(), 15) + BUDDHA)
        }
    }

}

File: iOS/PengChauFerry/HolidayRepository.swift
//
//  HolidayRepository.swift
//  PengChauFerry
//
//  Created by Lionello Lunesu on 7/3/2020.
//  Copyright © 2020 Lionello Lunesu. All rights reserved.
//

import Foundation

open class HolidayRepository {
    private let holidayDao = HolidayDao()
    private let BUDDHA = LocalDate.parse("2020-04-11T00:00:00Z")!

    open func getHoliday(day: LocalDate) -> Bool {
        return holidayDao.query(today: day)
    }

    open func setHoliday(day: LocalDate, isHoliday: Bool) {
        if isHoliday {
            holidayDao.insert(date: day)
        } else {
            holidayDao.delete(date: day)
        }
    }

    open func shouldRefresh() -> Bool {
        return !getHoliday(day: BUDDHA)
    }

    open func refresh(completion: () -> Void) {
        HongKongHolidayFetcher.fetch { holidays, _ in
            if holidays.count >= 15 {
                holidayDao.save(holidays: holidays + [BUDDHA])
            }
            completion()
        }
    }

}
